core/survey/_survey/surveyNodeDefs.js

```diff
@@ -129,7 +129,15 @@ export const getHierarchy = (filterFn = NodeDef.isEntity, includeAnalysis = fals
   const h = (array, nodeDef) => {
     const childDefs = [
       ...(NodeDef.isEntity(nodeDef) && !NodeDef.isVirtual(nodeDef)
-        ? R.pipe(getNodeDefChildren(nodeDef, includeAnalysis), R.filter(filterFn))(survey)
+        ? R.pipe(
+            getNodeDefChildren(nodeDef, includeAnalysis),
+            R.filter(filterFn),
+            // return virtual entities last (useful when inserting node defs into db)
```

Comment:
why is it useful?

Reply:
because the virtual entity definitions cannot be stored before the source entity definitions (parent entity foreign key) so it's useful to get them in this order

Reply:
This is not useful, this is very important! But It would more appropriate to sort the nodeDefs according to this rule when they are stored not in the model object because we can decide that the order can change in the future, not remembering why this rule existed.

Reply:
right, I'll move this sort to the test file where it's used.